blib/man3/Transaction.3
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Transaction 3"
.TH Transaction 3 "2010-08-09" "perl v5.10.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Transaction \- Perl extension for Transactions
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 3
\&        use Transaction;
\&        my $trn = new Transaction SERVER=>\*(Aqlocalhost\*(Aq, PORT=>50001;
\&        my $ret = $trn\->send(\*(Aqechotest\*(Aq, blah => \*(Aqfoo\*(Aq);
\&
\&        if ($ret\->param(\*(Aqerr\*(Aq) == 0) {
\&                print "$_=".$ret\->param($_) foreach $ret\->param();
\&        } else {
\&                die "Transaction err=". $ret\->param(\*(Aqerr\*(Aq);
\&        }
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Perl interface to send Transactions
.SH "METHODS"
.IX Header "METHODS"
.IP "\fBnew(...)\fR" 4
.IX Item "new(...)"
Creates a new Transaction
.IP "\fBparam(...)\fR" 4
.IX Item "param(...)"
.SH "AUTHOR"
.IX Header "AUTHOR"
Kiel R. Stirling, kiel@comcen.com.au
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIperl\fR\|(1).
.SH "POD ERRORS"
.IX Header "POD ERRORS"
Hey! \fBThe above document had some coding errors, which are explained below:\fR
.IP "Around line 167:" 4
.IX Item "Around line 167:"
\&'=item' outside of any '=over'
.IP "Around line 173:" 4
.IX Item "Around line 173:"
You forgot a '=back' before '=head1'


Changes
Revision history for Perl extension Transaction.

0.01  Sun May 25 16:21:44 2003
	- original version; created by h2xs 1.19



Transaction.pm
package Transaction;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;

@ISA = qw(Exporter AutoLoader);
#@EXPORT = qw();

$VERSION = '0.01';
$SIG{PIPE} = 'IGNORE';

my %errors = (
	1 => 'Username Invalid',
	2 => 'Incorrect password',
	3 => 'User not permitted from referring host',
	4 => 'Unknown trasaction',
	5 => 'User not permited to do transaction',
);

sub new {
	my ($self, %config) = @_;
	%config = map {uc $_=>$config{$_}} keys %config;
	return bless \%config, $self;
}

sub send {
	my ($self, $action, %args) = @_;

	my $trys = 0;
	delete $self->{_param} if exists $self->{_param};

	my $server;
	if (exists $self->{SSL} and $self->{SSL}) {
		use IO::Socket::SSL;
		$server = IO::Socket::SSL->new(
			PeerAddr => $self->{SERVER},
			PeerPort => $self->{PORT},
			Proto    => 'tcp',
			SSL_use_cert => 1,
			SSL_verify_mode => 0x01,
			SSL_passwd_cb => sub { return "testing" }
		) or die "unable to create socket: ", &IO::Socket::SSL::errstr, "\n";

	} else {
		#Open TCP port
		use IO::Socket;
		$server = IO::Socket::INET->new (
			PeerAddr => $self->{SERVER},	
			PeerPort	=> $self->{PORT},
			Type		=> SOCK_STREAM,
		) or die "Couldn't be a tcp server on port $self->{PORT} : $@\n";
	}
		
	print $server "ACTION=$action\n";
	print $server "USER=$self->{USER}\n";
	print $server "PASSWD=$self->{PASSWD}\n";
	print $server "$_=$args{$_}\n" foreach keys %args;
	
	shutdown $server, 1;
	
	$self->params($server, (exists $self->{SSL} and $self->{SSL}) ? '1' : 0);
	#print Dumper([$self]);
	shutdown $server, 2;
	#close $server or die $!;
	return $self;
}

sub _process_raw {
	my ($self, $raw) = @_;
	my ($key, $val) = split '=', $raw;
	return unless defined $key and defined $val;
	$val  =~ s/\r?\n$//;	
	if (exists $self->{_param}{$key}) {
		if (ref $self->{_param}{$key} eq 'ARRAY') {
			push @{$self->{_param}{$key}}, $val;
		} else {
			$self->{_param}{$key} = [$self->{_param}{$key}, $val]
		}	
	} else {
		$self->{_param}{$key} = $val;
	}
}

sub params {
	my ($self, $socket, $ssl) = @_;
	
	if ($ssl) {
		foreach ($socket->getlines) {
			$self->_process_raw($_);
		}
	} else {
	
		while (defined($_ = <$socket>) and not /^\n?\r$/) {
			$self->_process_raw($_);
		}
	}
	#print Dumper([$self->{_param}]);

	if (exists $self->{_param}{count}) {
		my %ret;
		for my $i (1 .. $self->{_param}{count}) {
			my %a;
			foreach my $key (grep /^$i\-/, $self->param()) {
				$key=~s/^$i\-//;
				$a{$key} = $self->param($&.$key); 

			}
			push @{$ret{$i}}, \%a
		}
		$ret{primary} = $self->{_param}{primary} if exists $self->{_param}{primary};
		$self->{_param} = \%ret;
	}	
}

sub param {
	my ($self, $param, $opr, $val) = @_;
	my $p = $self->{_param};
	
	if ($param and ref $p->{$param} eq 'ARRAY') {
		return $p->{$param};
	} elsif ($param and ref $p->{$param} eq 'HASH') {
		return $p->{$param};
	} elsif ($param) {
		return $p->{$param}
	} elsif ($opr and $opr eq '+') {
		$self->{_param}{$param} = $val;
	} else {
		if(wantarray()) {
			return keys %$p;
		}
	}
}

sub errstr {
	my $self = shift;
	return $errors{$self->param('err')};
}

1;

__END__

=head1 NAME

Transaction - Perl extension for Transactions

=head1 SYNOPSIS

	use Transaction;
	my $trn = new Transaction SERVER=>'localhost', PORT=>50001;
	my $ret = $trn->send('echotest', blah => 'foo');

	if ($ret->param('err') == 0) {
		print "$_=".$ret->param($_) foreach $ret->param();
	} else {
		die "Transaction err=". $ret->param('err');
	}
	
=head1 DESCRIPTION

Perl interface to send Transactions 

=head1 METHODS

=item B<new(...)>

Creates a new Transaction

=item B<param(...)>

=head1 AUTHOR

Kiel R. Stirling, kiel@comcen.com.au

=head1 SEE ALSO

perl(1).

=cut
